cktap/utils.py

```diff
@@ -51,15 +51,26 @@ def str2path(path):
     rv = []
 
     for i in path.split('/'):
-        if i == 'm': continue
-        if not i: continue      # trailing or duplicated slashes
+        if i == 'm':
+            continue
+        if not i:
+            # trailing or duplicated slashes
+            continue
 
         if i[-1] in "'phHP":
-            assert len(i) >= 2, i
-            here = int(i[:-1]) | HARDENED
+            if len(i) < 2:
+                raise ValueError(f"Malformed bip32 path component: {i}")
+            num = int(i[:-1])
+            if num >= HARDENED:
+                # cannot be more than (2 ** 31) - 1
+                raise ValueError(f"Hardened path component out of band: {i}")
+            here = num | HARDENED
         else:
-            here = int(i, 0)
-            assert 0 <= here < HARDENED, here
+            here = int(i)
+            if 0 > here or here >= HARDENED:
```

Comment:
`if 0 < here < HARDENED` (not `<=`)

Reply:
with a `not` inthere...

Reply:
if it is non-hardened, it should be in between 0 and 2**31-1 - so if it equals HARDENED it must fail

Reply:
it's a style thing: I want this to look like a range: `if not (0 <= here < HARDENED):`